File: solutions/fsharp/queen-attack/1/QueenAttack.fs
module QueenAttack

let create (position: int * int) =
    let x, y = position

    x >= 0 && x < 8 && y >=0 && y < 8

let canAttack (queen1: int * int) (queen2: int * int) =

    let q1x, q1y = queen1
    let q2x, q2y = queen2

    q1x = q2x || q1y = q2y || abs (q1x - q2x) = abs (q1y - q2y)

File: solutions/c/queen-attack/1/queen_attack.c
#include "queen_attack.h"
#include <stdlib.h>

attack_status_t can_attack(position_t queen_1, position_t queen_2) {
    if (queen_1.row >= 8 || queen_1.column >= 8 || queen_2.row >= 8 || queen_2.column >= 8)
        return INVALID_POSITION;

    int deltaX = queen_1.column - queen_2.column;
    int deltaY = queen_1.row - queen_2.row;

    if (deltaX == 0 && deltaY == 0)
        return INVALID_POSITION;

    if (deltaX == 0 || deltaY == 0 || abs(deltaX) == abs(deltaY))
        return CAN_ATTACK;

    return CAN_NOT_ATTACK;
}